### PR TITLE
0.6.15

### PR DESCRIPTION
**Pull request by yudilevi**

### CODE REVIEW

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 1]
```diff
@@ -31,43 +30,22 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
...
     mapping (ISmartToken => IERC20Token) private poolTokensToReserves;  // maps each pool token to its reserve
 
-    uint8 public amplificationFactor = 20;  // factor to use for conversion calculations (reduces slippage)
-    uint256 public externalRatePropagationTime = 1 hours;  // the time it takes for the external rate to fully take effect
-    uint256 public prevConversionTime;  // previous conversion time in seconds
-
-    // factors used in fee calculations
-    uint32 public lowFeeFactor = 200000;
-    uint32 public highFeeFactor = 800000;
+    Fraction public externalRate;           // conversion rate of 1 primary token in secondary tokens
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:33]: that's actually the external rate, not a conversion rate

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 2]
```diff
@@ -31,43 +30,22 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
...
+    uint256 public externalRateUpdateTime;  // last time the conversion rate was updated (in seconds)
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:34]: same here

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 3]
```diff
@@ -31,43 +30,22 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
...
     bool public maxStakedBalanceEnabled = true;
 
-    /**
-      * @dev triggered when the amplification factor is updated
-      *
-      * @param  _prevAmplificationFactor    previous amplification factor
-      * @param  _newAmplificationFactor     new amplification factor
-    */
-    event AmplificationFactorUpdate(uint8 _prevAmplificationFactor, uint8 _newAmplificationFactor);
+    uint32 public oracleDeviationFee = 0; // initial oracle deviation fee, represented in ppm
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:40]: it's not really the `initial` deviation fee, that property represents the oracle deviation fee and its default value is 0 (btw, I think the default can be 1%)

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 4]
```diff
@@ -556,15 +500,15 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
         IPoolTokensContainer(address(anchor)).mint(reservePoolToken, msg.sender, poolTokenAmount);
 
         // rebalance the pool's reserve weights
-        rebalance(rate);
+        rebalance();
 
-        // dispatch the LiquidityAdded event
+        // dispatch the liquidity-added event
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:505]: why not use the real event names here?

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 5]
```diff
@@ -620,17 +561,17 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
             safeTransfer(reserveToken, msg.sender, reserveAmount);
 
         // rebalance the pool's reserve weights
-        rebalance(rate);
+        rebalance();
 
         uint256 newPoolTokenSupply = initialPoolSupply.sub(_amount);
 
-        // dispatch the LiquidityRemoved event
+        // dispatch the liquidity-removed event
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:568]: why not use the real event names here?

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 6]
```diff
@@ -873,141 +674,21 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
...
-        // if one of the balances is 0, return the external rate
-        if (reserves[primaryReserveToken].balance == 0 || reserves[secondaryReserveToken].balance == 0) {
-            Fraction memory externalRate;
-            (externalRate.n, externalRate.d) = priceOracle.latestRate(primaryReserveToken, secondaryReserveToken);
-            return externalRate;
-        }
-
-        // return the rate based on the current rate
-        return tokensRate(primaryReserveToken, secondaryReserveToken, 0, 0);
+    function _effectiveTokensRate() private view returns (Fraction memory) {
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:681]: we only use that function for converting between uint256,uint256 and Fraction, I wonder if we can make cleaner - maybe a utility function that does that?

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 7]
```diff
@@ -1019,101 +700,62 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
...
-      *
-      * @return effective rate of 1 primary token in secondary tokens
-    */
-    function rateFromPrimaryWeight(uint32 _primaryReserveWeight) private view returns (Fraction memory) {
-        uint256 n = stakedBalances[secondaryReserveToken].mul(_primaryReserveWeight);
-        uint256 d = stakedBalances[primaryReserveToken].mul(inverseWeight(_primaryReserveWeight));
-        (n, d) = reducedRatio(n, d, MAX_RATE_FACTOR_LOWER_BOUND);
-        return Fraction(n, d);
+        // get the tokens amplified balances
+        uint256 primaryBalance = reserveAmplifiedBalance(primaryReserveToken);
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:719]: pretty sure we can call `amplifiedBalance` here to skip the `validReserve` check in `reserveAmplifiedBalance`

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol [hunk 8]
```diff
@@ -380,25 +336,31 @@ contract LiquidityPoolV2Converter is LiquidityPoolConverter {
...
-        if (_targetToken == primaryReserveToken) {
-            sourceTokenWeight = inverseWeight(sourceTokenWeight);
-            externalSourceTokenWeight = inverseWeight(externalSourceTokenWeight);
+        uint32 sourceTokenWeight;
+        uint32 targetTokenWeight;
+
+        // if the rate was already checked in this block, use the current weights; otherwise, get the new weights
+        if (externalRateUpdateTime == time()) {
+            sourceTokenWeight = reserves[_sourceToken].weight;
+            targetTokenWeight = reserves[_targetToken].weight;
```
Comment (yudilevi) [inline solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:345]: can use PPM_RESOLUTION - sourceTokenWeight here to save some gas

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 3 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:40]: But we have two oracles, so 2%?

Comment (yudilevi) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:40]: no I don't think it's related, 1% is a good start as the entire deviation.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 6 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:681]: I don't understand this comment.

Comment (yudilevi) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:681]: I thought we could create a utility function that receives 2 uint256 values and returns a newly constructed Faction struct, and then get rid of `_effectiveTokensRate`, but some to think of it, I don't believe you can pass a tuple return value to function as argument

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 1 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:33]: Fixed.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 2 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:34]: Fixed.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 4 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:505]: Fixed.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 5 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:568]: Fixed.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 7 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:719]: Fixed in all places; please verify.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 8 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:345]: Fixed in all places; please verify.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 6 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:681]: I still don't understand.
We'd still need to call `priceOracle.latestRate(primaryReserveToken, secondaryReserveToken)`.
So what exactly are you suggesting (ideally, assuming that we don't have a compiler limitation)?

Comment (yudilevi) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol:681]: was just looking for ways to remove that function that basically does nothing.

File: solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol (hunk 3 above)
Comment (barakman) [reply solidity/contracts/converter/types/liquidity-pool-v2/LiquidityPoolV2Converter.sol]: Fixed.

**Pull request merged**